Assets/Character/AaronTestAssets/PlayerMovement.cs

```diff
@@ -6,79 +6,108 @@
-                Flip(horizontal);
+            float horizontal = (Input.GetAxis("Horizontal"));
+
+            //position based
+            transform.position += new Vector3(horizontal, 0, 0) * move_speed * Time.deltaTime;
+
+            //add force
+            //rigidbody.AddForce(new Vector2(horizontal * speed, 0), ForceMode2D.Force);
+
+                if (animator.GetBool("OnGround"))
+                {
+                    if (horizontal != 0)
+                    {
+                        animator.Play("Walk");
+                        Flip(horizontal);
+                    }
+
+                    else
+                    {
+                        animator.Play("Idle");
+                    }
+
+                    if (Input.GetKeyDown("space"))
+                    {
+                        animator.SetBool("OnGround", false);
+                        animator.Play("Jump");
+                        rigidbody.AddForce(new Vector2(0, jump_power), ForceMode2D.Impulse);
+                    }
+
+                    if (Input.GetKeyDown(KeyCode.X))
+                    {
+                        animator.SetBool("OnGround", false);
+                        animator.SetBool("Rolling", true);
+                        timer = Time.time;
+                    }
+
+                }
             }
-            else
+        else 
+        {
+            float timePassed = Time.time-timer;
+            print(timePassed);
+            if (timePassed < roll_time)
             {
-                animator.Play("Idle");
+                animator.Play("Roll");
+                Flip(directionFacing);
+                rigidbody.AddForce(new Vector2(roll_speed * directionFacing, 0), ForceMode2D.Impulse);
+                print("adding force");
             }
-            if (Input.GetKeyDown("space"))
+            else
             {
-                animator.SetBool("OnGround", false);
-                animator.Play("Jump");
-                rigidbody.AddForce(new Vector2(0, jump), ForceMode2D.Impulse);
+                animator.SetBool("OnGround", true);
+                animator.SetBool("Rolling", false);
             }
         }
-
-
     }
 
     void Flip(float horizontal)
     {
-        if ((isFacingRight && horizontal < 0) || (!isFacingRight && horizontal > 0))
+        if ((directionFacing == 1 && horizontal < 0) || (directionFacing == -1 && horizontal > 0))
         {
-            Vector3 scale = transform.localScale;
             scale.x *= -1;
             transform.localScale = scale;
 
-            if (horizontal > 0)
-            {
-                isFacingRight = true;
-            }
+            directionFacing = (int)(horizontal / Mathf.Abs(horizontal));
 
-            if (horizontal < 0)
-            {
-                isFacingRight = false;
-            }
         }   
     }
 
 
     void OnCollisionEnter2D(Collision2D coll)
     {
-
+        //maybe add a bit of space above the ground to be considered OnGround
```

Comment:
or possibly queue the input. Input queue needs to be implemented for combos anyway.